docs/design/uplink-contexts.md

```diff
@@ -0,0 +1,103 @@
+    EncryptionKey   []byte
+}
+```
+
+<blockquote>
+For concreteness, the following is an example UX around uplink with contexts. It is expected to evolve, and is just used to help with understanding.
+</blockquote>
+
+When setting up uplink for the first time, a default context is created, and the command will prompt you for some credentials. For example,
+
+```
+$ uplink setup
+A new "default" context has been created.
+
+Please respond with the type of information you have:
+1. API key from a satellite
+2. Exported context
+> 1
+
+API Key: <user pastes key>
+Satellite URL: <user puts in or picks satellite like now>
+```
+
+After a context has been created, bucket information must be added. For example,
+
+```
+$ uplink import [optionally specify a context with --context=default]
+
+Please respond with the type of information you have:
+1. Bucket with root passphrase
+2. Bucket share
+> 1
+
+Bucket name: <user inputs bucket name>
+Passphrase: <user inputs passphrase>
+
+$ uplink import
+Please respond with the type of information you have:restricted
+1. Bucket with root passphrase
+2. Bucket share
+> 2
+
+Bucket share: <user pastes shared bucket (RevealedPath)>
+```
+
+When a client attempts to interact with a bucket, it finds the latest added matching unencrypted path and appends the rest of the path, encrypted with the key the revealed path specifies, to the revealed path's encrypted path. For example, if one had this set of `BucketShare` objects:
+
+```go
+paths := []BucketShare{
+    {Bucket: "x", EncryptedPath: "/e1/e2/e3", UnecryptedPath: "/a/b/c", EncryptionKey: "key1"},
+    {Bucket: "x", EncryptedPath: "/e1",       UnecryptedPath: "/a",     EncryptionKey: "key2"},
+    {Bucket: "x", EncryptedPath: "/e4",       UnecryptedPath: "/f",     EncryptionKey: "key3"},
+}
+```
+
+Then if one asked for `sj://x/a/d`, the 2nd path would be the longest match and `"key2"` would be used to encrypt `d` and it would be appened to `"/e1"`. Similarly, if one asked for `sj://x/a/b/c/f`, then `"key2"` would be used to encrypt `b/c/f` and it would be appened to `"/e1"`. If one asked for `sj://x/f/g`, then `"key3"` would be used. `"key1"` would never be used as it is subsumed by the entry with `"key2"`.
+
+## Rationale
+
+This proposal is not a radical departure of the current method, just an extension to allowing a collection of paths to compose a bucket at any unencrypted spot. Thus, it shoudn't require any changes on the server side, reducing risk and allowing for earlier shipping.
+
+We could allow for "longest" matching. For example, in the above set of revealed paths one could use `"key1"` for `sj://x/a/b/c/d`. This begs a question about consistency. In principle be able to derive `"key1"` from the information provided by `"key2"`. If they disagree, an error could be thrown. This design takes the position that, these issues are trivial because it is very unlikely that such a share would be constructed, and even if it were, it would disagree.
```

Comment:
It probably ought to be mentioned in this doc that the "unencrypted path" part of a bucket share does not need to match the original path from the project root. Allowing for this shouldn't break anything and gives more flexibility to the users. Furthermore, I don't think anything in our system prohibits somebody from using some arbitrary unrelated encryption key for files under some path prefix, if they feel like it. Given that, it will not be possible in the general case to derive key1 from key2, and it would not be particularly useful to do so anyway.

Reply:
Indeed. Thinking over that more, that makes me think that doing "longest" matching with ties broken by later entries would be the best, because it allows overrides.